File: src/ch04/solution_2.py
from typing import List

from rich import print


def tribonacci(n: int, memo: List[int]) -> int:
    if memo[n] != -1:
        return memo[n]
    memo[n] = tribonacci(n=n-1, memo=memo) + tribonacci(n=n-2, memo=memo) + tribonacci(n=n-3, memo=memo)
    return memo[n]


if __name__ == '__main__':
    n: int = 10
    memo: List[int] = [-1] * (n+1)
    memo[0], memo[1], memo[2] = 0, 0, 1
    result: int = tribonacci(n=n, memo=memo)

    print(result)

File: src/ch04/solution_2.rs
fn tribonacci(n: usize, memo: &mut Vec<i32>) -> i32 {
    if memo[n] != -1 {
        return memo[n];
    }
    memo[n] = tribonacci(n-1, memo) + tribonacci(n-2, memo) + tribonacci(n-3, memo);
    memo[n]
}


fn main() {
    let n: usize = 10;
    let mut memo: Vec<i32> = vec![-1; n+1];
    memo[0] = 0;
    memo[1] = 0;
    memo[2] = 1;
    let result: i32 = tribonacci(n, &mut memo);

    println!("{}", result);
}